### DISCUSSION TITLE
Class not found: jakarta/servlet/ServletContextListener with OpenGrok 1.12.15 and Tomcat 9

### DISCUSSION THREAD

**Discussion by randygalbraith** [oracle/opengrok#4416] [General]
Hi Everyone. I am seeking assistance getting OpenGrok running.

I am unable to deploy source.war (from OpenGrok 1.12.15) to Tomcat 9.

I copy source.war to webapps/opengrok.war. Tomcat then expands the war file. This results in a failed deploy.

From localhost.2023-10-02.log:
java.lang.NoClassDefFoundError: jakarta/servlet/ServletContextListener

Version details from the Tomcat manager:

Tomcat Version: Apache Tomcat/9.0.80
JVM Version: 11.0.20.1+1
JVM Vendor: Eclipse Adoptium
OS Name: Linux
OS Version: 4.18.0-500.el8.x86_64
OS Architecture: amd64

The OS name from /etc/os-release is: CentOS Stream 8.

Here is the  <Connector> info from server.xml:

  <Service name="Catalina">
 
    <Connector port              = "xxxxx"
               protocol          = "org.apache.coyote.http11.Http11NioProtocol"
               maxThreads        = "150"
               SSLEnabled        = "true"
               maxParameterCount = "1000" >
 
      <UpgradeProtocol className =
                       "org.apache.coyote.http2.Http2Protocol" />
 
        <SSLHostConfig protocols = "TLSv1.2" >
          <Certificate
              certificateKeyFile      = "<path-to-key>"
              certificateFile         = "<path-to-cert>"
              certificateChainFile    = "<path-to-cert>"
              certificateKeyPassword  = "<valid-password>"
              type                    = "RSA" />
        </SSLHostConfig>

    </Connector>

Our team has used OpenGrok for years. This is an effort to upgrade to the latest version of OpenGrok.

Kind regards, -Randy

**Comment by vladak** [comment]
You need at least Tomcat 10.0.

**Reply by randygalbraith** [reply]
Thanks vladak. I feel rather silly now given the front page says: "Tomcat (10.x or later)". Alas, I've been working a number of upgrades and somehow missed that. Alas, for the moment, I must stick with Tomcat 9. I'll now look for the latest version of OpenGrok that supports Tomcat 9.

I see 1.6.0 says "OpenGrok now requires Tomcat 10" so I'll try 1.5.12.

Kind regards, -RAndy

**Reply by vladak** [reply]
No worries. I wish there was a way how to make this more robust, somehow making the `.war` file indicate the needed app server version/level.

**Reply by randygalbraith** [reply]
Hi vladak. I downloaded 1.5.12 and it deployed without problem. I was happy to see opengrok 1.3.4 makes a reference to Universal Ctags. So that ctags upgrade I'll be able to move forward with. Again, thanks, so much! Kind regards, -Randy
